hardware/RelayNet.cpp

```diff
@@ -0,0 +1,748 @@
+	m_bIsStarted		= false;
+
+	m_stoprequested=false;
+	m_bDoRestart=false;
+
+	//force connect the next first time
+	m_retrycntr=RETRY_DELAY;
+
+
+	if (m_input_count || m_relay_count)
+	{
+		m_thread = boost::shared_ptr<boost::thread>(new boost::thread(boost::bind(&RelayNet::Do_Work, this)));
+	}
+
+	if (m_thread != NULL)
+	{
+		bOk = true;
+		m_bIsStarted=true;
+	}
+
+	return bOk;
+}
+
+//===========================================================================
+
+bool RelayNet::StopHardware()
+{
+	m_stoprequested = true;
+
+	if (isConnected())
+	{
+		disconnect();
+	}
+
+	try 
+	{
+		if (m_thread)
+		{
+			m_thread->join();
+		}
+	}
+	catch (...)
+	{
+	}
+
+	m_bIsStarted = false;
+
+	_log.Log(LOG_STATUS, "RelayNet: Relay Module disconnected %s", m_szIPAddress.c_str());
+
+	return true;
+}
+
+//===========================================================================
+
+bool RelayNet::WriteToHardware(const char *pdata, const unsigned char length)
+{
+	bool bOk = true;
+
+#if RELAYNET_USE_HTTP
+
+	bOk = WriteToHardwareHttp(pdata, length);
+
+#else
+
+	bOk = WriteToHardwareTcp(pdata, length);
+
+#endif
+
+	return bOk;
+}
+
+//===========================================================================
+
+void RelayNet::Do_Work()
+{
+	bool bFirstTime=true;
+	int sec_counter = 0;
+
+	/*  Init  */
+	Init();
+
+	if (m_poll_inputs || m_poll_relays)
+	{
+		_log.Log(LOG_STATUS, "RelayNet: %d-second poller started (%s)", m_poll_interval, m_szIPAddress.c_str());
+	}
+
+	while (!m_stoprequested)
+	{
+		/*  One second sleep  */
+		sleep_seconds(1);
+		sec_counter++;
+
+		/*  Heartbeat maintenance  */
+		if (sec_counter  % 10 == 0)
+		{
+			m_LastHeartbeat = mytime(NULL);
+		}
+
+		/*  Connection maintenance  */
+		if (bFirstTime)
+		{
+			bFirstTime = false;
+			connect(m_szIPAddress,m_usIPPort);
+		}
+		else
+		{
+			if ((m_bDoRestart) && (sec_counter % 30 == 0))
+			{
+				connect(m_szIPAddress,m_usIPPort);
+			}
+			update();
+		}
+
+		/*  Prevent disconnect request by Relay Module  */
+		if ((sec_counter % KEEP_ALIVE_INTERVAL == 0) &&
+			((m_poll_interval > KEEP_ALIVE_INTERVAL) || (!m_poll_inputs && !m_poll_relays)))
+		{
+			KeepConnectionAlive();
+		}
+
+		/*  Update relay module status when poll interval has expired  */
+		if ((m_poll_inputs || m_poll_relays) && (sec_counter % m_poll_interval == 0))
+		{
+			TcpRequestRelaycardDump();
+		}
+	}
+
+	/*  Done  */
+	if (m_poll_inputs || m_poll_relays)
+	{
+		_log.Log(LOG_STATUS, "RelayNet: %d-second poller stopped (%s)", m_poll_interval, m_szIPAddress.c_str());
+	}
+}
+
+//===========================================================================
+
+void RelayNet::Init()
+{
+	BYTE	id1 = 0x03;
+	BYTE	id2 = 0x0E;
+	BYTE 	id3 = 0x0E;
+	BYTE 	id4 = m_HwdID & 0xFF;
+
+	/* 	Prepare packet for LIGHTING2 relay status packet  */
+	memset(&Packet, 0, sizeof(RBUF));
+
+	if (m_HwdID > 0xFF)
+	{
+		id3 = (m_HwdID >> 8) && 0xFF;
+	}
+
+	if (m_HwdID > 0xFFFF)
+	{
+		id3 = (m_HwdID >> 16) && 0xFF;
```

Comment:
I've been reviewing compiler warnings, and by judging from the code you should use a bitwise operation not a logical and. Change && to & (same for id3 above).
Also, shouldn't this be id2 (hence the >> 16)?
Why not the same logic for id1 also?